### ISSUE TITLE
Crash when using GDAL compiled with clang and libc++ in Unreal Engine

### ISSUE THREAD

**Issue by LandscapeCombinator** [OSGeo/gdal#9738]
### What is the bug?

Hello, I'm trying to use GDAL in a plugin I'm developing for Unreal Engine 5.4. The plugin works well under Windows using GDAL binaries downloaded from conda and I'm now trying to port the plugin to Linux. I have tried taking GDAL binaries for Linux from conda or by building them using conan. The plugin compiles fine, but I'm getting a crash when trying to open a tif file with  `GDALOpen(file, GA_READONLY)`, which opens fine using the system's gdalinfo.

This is Unreal Engine's log when crashing, I think the top part is related to UE and not to the actual error, whose stack trace ends with `_tiffCloseProc`.

```
Signal 11 caught.
Malloc Size=262146 LargeMemoryPoolOffset=262162 
CommonUnixCrashHandler: Signal=11
LogCore: === Critical error: ===
Unhandled Exception: SIGSEGV: unaligned memory access (SIMD vectors?)

LogCore: Fatal error!

0x00007fdf0a16f4bc libUnrealEditor-Module.so!_tiffCloseProc(void*)()
0x00007fe038a5d39b libUnrealEditor-Module.so!_TIFFSeekOK(+0x1a)
0x00007fe038a645e0 libUnrealEditor-Module.so!TIFFFetchDirectory()
0x00007fe038a62c5d libUnrealEditor-Module.so!TIFFReadDirectory(+0x12c)
0x00007fdf09fbbcaa libUnrealEditor-Module.so!TIFFClientOpenExt(+0x6f9)
0x00007fdf0a16f114 libUnrealEditor-Module.so!VSI_TIFFOpen_common(GDALTiffHandle*, char const*)()
0x00007fdf0a174c87 libUnrealEditor-Module.so!GTiffDataset::Open(GDALOpenInfo*)()
0x00007fdf0b05fda1 libUnrealEditor-Module.so!GDALDriver::Open(GDALOpenInfo*, bool)(+0x20)
0x00007fdf0b06c949 libUnrealEditor-Module.so!GDALOpenEx(+0x7e8)
0x00007fdf09905d2e libUnrealEditor-Module.so!MyFile::MyFunction() [MyFile.cpp:442]
```

In conan, I have also tried building shared libraries, or trying another recipe such as gdal 3.7.3, but I'm getting the same error. I also tried the shared libraries from conda, which give a similar error.
I also tried using a debug build in conan for gdal 3.8.3, and this time the program does not crash, but returns this error:

```
: "�z)�": Bad mode.
ERROR 4: `/home/user/file.tif' not recognized as a supported file format.
```

This file opens fine when directly using the `gdalinfo` binary from conda

### Steps to reproduce the issue

Currently, reproducing the issue is not practical because it involves downloading Unreal Engine. I can try to make a self-contained reproduction package if necessary.

### Versions and provenance

Debian GNU/Linux 11 (bullseye)
clang version 16.0.6 (from UE 5.4)
libc++ from UE 5.4
Built gdal 3.8.3 using conan (also tried 3.8.5 from conda, and 3.7.3 with conan)

I don't know how the conda binaries are built for gdal, so the `clang` and `libc++` versions only apply for the conan builds of GDAL.

**Comment by rouault** [comment]
> I don't know how the conda binaries are built for gdal,

using gcc I believe

It is likely you need to build GDAL by yourself using the same compiler as Unreal Engine.

It looks very much like an integration issue. Nothing that we can do on the GDAL source side. Closing as such

**Comment by LandscapeCombinator** [comment]
> > I don't know how the conda binaries are built for gdal,
> 
> using gcc I believe
> 
> It is likely you need to build GDAL by yourself using the same compiler as Unreal Engine.
> 
> It looks very much like an integration issue. Nothing that we can do on the GDAL source side. Closing as such

Thanks for the quick answer. This is what I have done for my attempt with conan. I built the binaries myself using conan, with the clang compiled shipped from UE, and the libc++ shipped from UE. (The attempt with the conda binaries is a different attempt than the conan attempts, which lead to the same error.)

**Comment by rouault** [comment]
It might be a mix of libraries. GDAL links against libtiff. But perhaps Unreal Engine does link against libtiff too. Are the 2 libtiff the same one... This is a "classic" clash in complex environments.

**Comment by LandscapeCombinator** [comment]
Oh you're right, there is a libtiff inside Unreal Engine, I'll try that, thank you!

**Comment by LandscapeCombinator** [comment]
This worked! Thanks a lot for the help, which was very quick and to the point!

I'm having another issue, but I'm not sure if it's again related to a mismatch between statically linked libraries or library versions. I can open another issue in the proj repository if this is more appropriate. I can also try to make a reproduction package without UE if this helps.

`GDALTranslate` and `GDALWarp` worked well, but OGR transformations are not able to find basic codes such as "EPSG:4326". I'm setting `GDAL_DATA`, `PROJ_DATA`, `PROJ_LIB` using `CPLSetConfigOption()` to the correct "share" directories obtained from the conan build. The gdal directory contains csv files, epsg.wkt, etc., while the proj directory contains proj.db and other files.

On Windows (using conda binaries), this was enough for my code to work, but now I'm getting an error when trying to create a transformation. The code which fails does the following. It creates two `OGRSpatialReference` and sets them both to "EPSG:4326" using `SetFromUserInput` (both invocations return `OGRERR_NONE`). Then, `OGRCreateCoordinateTransformation` is called on these two `OGRSpatialReference`'s but returns nil with these errors:


```
ERROR 1: PROJ: proj_crs_get_coordinate_system: Object is not a SingleCRS
ERROR 1: PROJ: proj_crs_get_coordinate_system: Object is not a SingleCRS
ERROR 1: PROJ: proj_crs_get_datum: Object is not a SingleCRS
ERROR 1: PROJ: proj_crs_get_datum_ensemble: Object is not a SingleCRS
ERROR 1: PROJ: proj_crs_get_coordinate_system: Object is not a SingleCRS
ERROR 1: PROJ: proj_crs_get_datum: Object is not a SingleCRS
ERROR 1: PROJ: proj_crs_get_datum_ensemble: Object is not a SingleCRS
ERROR 1: PROJ: proj_crs_get_coordinate_system: Object is not a SingleCRS
ERROR 1: PROJ: proj_create: unrecognized format / unknown name
ERROR 1: PROJ: proj_create: unrecognized format / unknown name
```

**Comment by LandscapeCombinator** [comment]
> that should typically be something/share/gdal for GDAL_DATA and something/share/proj for PROJ_DATA

yes that's correct

> Creating a reproducer outside of UE would be desirable, but I feel this is "just" an issue with setting the environment right. You might also try setting the variables as environment variables, although that should work with CPLSetConfigOption()

Just tried with environment variables too, but same error, I'll try to make a reproducer, thanks again for the help!

**Comment by LandscapeCombinator** [comment]
I finally found the issue! (when trying to make a reproducer that wouldn't fail)
I have different "modules" in my Unreal Engine project, and I needed to call `CPLSetConfigOption` and  `OGRSpatialReference` from all the modules where I used `OGRCreateCoordinateTransformation`. I was also making a call `SetFromUserInput` in one module (which was succeeding), and `OGRCreateCoordinateTransformation` in another module, which doesn't work. This is maybe because I'm statically linking the libraries. (while on Windows I wasn't having these issues because I was using shared libraries).

Thank you for pointing me the right direction!
